plugins/repository-encrypted/src/main/java/org/elasticsearch/repositories/encrypted/EncryptedRepository.java

```diff
@@ -0,0 +1,308 @@
+            for (String pathComponent : path) {
+                encryptionMetadataBlobPath = encryptionMetadataBlobPath.add(pathComponent);
+            }
+            return new EncryptedBlobContainerDecorator(this.delegatedBlobStore.blobContainer(path),
+                    this.delegatedBlobStore.blobContainer(encryptionMetadataBlobPath), this.masterSecretKey);
+        }
+    }
+
+    private static class EncryptedBlobContainerDecorator implements BlobContainer {
+
+        private final BlobContainer delegatedBlobContainer;
+        private final BlobContainer encryptionMetadataBlobContainer;
+        private final SecretKey masterSecretKey;
+
+        EncryptedBlobContainerDecorator(BlobContainer delegatedBlobContainer, BlobContainer encryptionMetadataBlobContainer,
+                SecretKey masterSecretKey) {
+            this.delegatedBlobContainer = delegatedBlobContainer;
+            this.encryptionMetadataBlobContainer = encryptionMetadataBlobContainer;
+            this.masterSecretKey = masterSecretKey;
+        }
+
+        @Override
+        public BlobPath path() {
+            return this.delegatedBlobContainer.path();
+        }
+
+        @Override
+        public InputStream readBlob(String blobName) throws IOException {
+            final BytesReference dataDecryptionKeyBytes = Streams.readFully(this.encryptionMetadataBlobContainer.readBlob(blobName));
+            try {
+                SecretKey dataDecryptionKey = unwrapKey(BytesReference.toBytes(dataDecryptionKeyBytes), this.masterSecretKey);
+                Cipher cipher = Cipher.getInstance(ENCRYPTION_MODE, BC_PROV);
+                cipher.init(Cipher.DECRYPT_MODE, dataDecryptionKey, ivParameterSpec);
+                return new CipherInputStream(this.delegatedBlobContainer.readBlob(blobName), cipher);
+            } catch (InvalidKeyException | NoSuchAlgorithmException | NoSuchPaddingException | InvalidAlgorithmParameterException e) {
+                throw new IOException(e);
+            }
+        }
+
+        @Override
+        public void writeBlob(String blobName, InputStream inputStream, long blobSize, boolean failIfAlreadyExists) throws IOException {
+            try {
+                SecretKey dataEncryptionKey = generateRandomSecretKey();
+                byte[] wrappedDataEncryptionKey = wrapKey(dataEncryptionKey, this.masterSecretKey);
+                try (InputStream stream = new ByteArrayInputStream(wrappedDataEncryptionKey)) {
+                    this.encryptionMetadataBlobContainer.writeBlob(blobName, stream, wrappedDataEncryptionKey.length, failIfAlreadyExists);
+                }
+                Cipher cipher = Cipher.getInstance(ENCRYPTION_MODE, BC_PROV);
+                cipher.init(Cipher.ENCRYPT_MODE, dataEncryptionKey, ivParameterSpec);
+                this.delegatedBlobContainer.writeBlob(blobName, new CipherInputStream(inputStream, cipher), blobSize + GCM_TAG_BYTES_LENGTH,
```

Comment:
CipherInputStream does not support `marking` (i.e. `CipherInputStream.markSupported()` returns false) which means that failed requests to S3 / GCS can't be auto-retried.

Reply:
Urgh good find ... I guess we could create our own cipher stream that supports it, but then we'd either have to buffer the encrypted data to disk or reencrypt the whole thing on retry right? (maybe there's a better way that I don't see?)

Reply:
Is those repositories (blob stores) only require reseting back to the beginning of the stream (to retry the whole request) then we could probably handle it with reencryption.
But if we want to `mark` an arbitrary point in the stream and `reset` to it, then re-encryption probably isn't going to be an option because you can't reset a `Cipher` to an artibrary point, and that means any checksums will be thrown out because the Cipher will think it has processed more data than was actually written to the blob store.

Reply:
> But if we want to mark an arbitrary point in the stream and reset to it, then re-encryption probably isn't going to be an option because you can't reset a Cipher to an artibrary point, and that means any checksums will be thrown out because the Cipher will think it has processed more data than was actually written to the blob store.

My thinking was to simply reset back to `0` to get a fresh Cipher, but then just dump the bytes up to whatever point `x` we actually want to reset to and only start producing the bytes starting from `x` to have a CPU-expensive by IO-cheap way of implementing resetting to an arbitrary point. Wouldn't that work?